src/poetry/installation/executor.py

```diff
@@ -148,6 +150,19 @@ def execute(self, operations: list[Operation]) -> int:
                 if self._shutdown:
                     break
 
+                # It's not safe to clone the same git repository multiple times in parallel
+                if operation.package.source_type == "git" and not operation.skipped:
+                    # If this is the first operation cloning this repository,
+                    # allow it to be executed in parallel
+                    if operation.package.source_url not in git_repositories_to_clone:
```

Comment:
Does this work on forked vs non-forked?
https://github.com/repo1/code
https://github.com/repo2/code

Reply:
Good catch, it does not! The race condition stems from having multiple installs that try to clone to the same local directory. The name for this directory is created [here](https://github.com/python-poetry/poetry/blob/7181557d01cc8c6e4eb15014a1e736fcfaea3a15/src/poetry/vcs/git/backend.py#L426).

In your example, both repositories would get the same `target`: `<source_root>/code`.

But as you point out, I would not catch this in my checks for identical source_urls, since the forks have different urls.

Reply:
Or 3. fix the race condition ?

From the little I understand, it appears that the root cause is concurrent execution of `poetry.vcs.git.Git.clone()` in the same physical directory. Neither the `git` executable, nor dulwich support concurrent execution in the same directory. Does this sound right?

What if we prevent concurrent git execution in the same directory? We could use `threading.Lock`.

In `vcs/git/backend.py` this would look like:
```python
from collections import defaultdict
from threading import Lock

_clone_lock: dict[Path, Lock] = defaultdict(Lock)

@classmethod
def clone(
    cls,
    url: str,
    name: str | None = None,
    branch: str | None = None,
    tag: str | None = None,
    revision: str | None = None,
    source_root: Path | None = None,
    clean: bool = False,
) -> Repo:
    source_root = source_root or cls.get_default_source_root()
    source_root.mkdir(parents=True, exist_ok=True)

    name = name or cls.get_name_from_source_url(url=url)
    target = source_root / name
    refspec = GitRefSpec(branch=branch, revision=revision, tag=tag)

    # Prohibit concurrent clones in the same target Path
    with cls._clone_lock[target]:
      # rest of clone follows, only the indentation changes
```
This will deadlock if `clone()` calls back into `clone()` _using the same directory_. And although recursive calls do happen via submodules, it doesn't look like they use the same directory.

I tried this out on the sample repo in https://github.com/python-poetry/poetry/issues/6958 and it seems to fix the issue, but I'm not familiar enough with the poetry code to know if this would be a viable approach.

Reply:
Nice! Fixing the problem at it's root would of course be great.

An unfortunate functional drawback would be that when multiple installs from the same repository are executed in parallel, they would all execute past the [cache lookup](https://github.com/python-poetry/poetry/blob/258ac696efb4c4c6abfc3352036a1bd1f7135d03/src/poetry/installation/executor.py#L620), and then proceed to wait right before the clone.

I have not dug deep into this, but it seems like this would cause the repository to be downloaded multiple times instead of just once. With the solution in this PR, it's only downloaded once.

Reply:
Good point. My suggestion in https://github.com/python-poetry/poetry/pull/9658#discussion_r1751154938 could potentially work, but would cause a performance hit for sure.

So how about this: git operations with the same url name (from `Git.get_name_from_source_url()`) might conflict, because they're cloned in `<source_root>/src/<url name>`. So we serialize all git operations which share an url name.

For example, on top of poetry main:
```diff
diff --git a/src/poetry/installation/executor.py b/src/poetry/installation/executor.py
index 196cdd5a..ff439d50 100644
--- a/src/poetry/installation/executor.py
+++ b/src/poetry/installation/executor.py
@@ -32,6 +32,7 @@ from poetry.utils.helpers import pluralize
 from poetry.utils.helpers import remove_directory
 from poetry.utils.isolated_build import IsolatedBuildError
 from poetry.utils.isolated_build import IsolatedBuildInstallError
+from poetry.vcs.git import Git
 
 
 if TYPE_CHECKING:
@@ -144,6 +145,7 @@ class Executor:
         for _, group in groups:
             tasks = []
             serial_operations = []
+            git_operations = []
             for operation in group:
                 if self._shutdown:
                     break
@@ -162,8 +164,32 @@ class Executor:
                     serial_operations.append(operation)
                     continue
 
+                # Defer git operations, they might need to be serialized
+                if operation.package.source_type == "git" and not operation.skipped:
+                    git_operations.append(operation)
+                    continue
+
                 tasks.append(self._executor.submit(self._execute_operation, operation))
 
+            def url_name_from_operation(operation: Operation) -> str | None:
+                if url := operation.package.repository_url:
+                    return Git.get_name_from_source_url(url)
+                return None
+
+            # git operations on the same url name end up being cloned in the same directory,
+            # so these must execute serially or they'll interfere with each other.
+            for _, git_ops_group in itertools.groupby(
+                git_operations,
+                key=url_name_from_operation,
+            ):
+                ops = list(git_ops_group)
+
+                def _serialize(ops: list[Operation] = ops) -> None:
+                    for op in ops:
+                        self._execute_operation(op)
+
+                tasks.append(self._executor.submit(_serialize))
+
             try:
                 wait(tasks)
```

Reply:
I'm leaning towards ensuring that forked repositories (actually any repository that happen to have the same name) are not cloned into the same target directory. I'd have to examine if there are any implicit dependencies on the name of the target folder though.

If that route seems acceptable, I would also consider implementing it in a separate PR, since it could be viewed as a separate fix.

What do you think?

Reply:
A separate PR sounds right to me.

I am wondering what the use case behind this issue is? If you have two forks with the same name the two packages that are built from the forks are likely to have the same name, right? If you have two packages with the same name from different repos you will get a conflict because you can only install one package with a specific name into an environment.